ffi/diplomat/src/errors.rs

```diff
@@ -0,0 +1,147 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use self::ffi::ICU4XError;
+use core::fmt;
+use fixed_decimal::Error as DecimalError;
+use icu_decimal::FixedDecimalFormatError;
+use icu_plurals::PluralRulesError;
+use icu_properties::PropertiesError;
+use icu_provider::{DataError, DataErrorKind};
+
+#[diplomat::bridge]
+pub mod ffi {
+    use alloc::boxed::Box;
+
+    #[repr(C)]
+    /// A common enum for errors that ICU4X may return, organized by API
+    ///
+    /// The error names are stable and can be checked against as strings in the JS API
+    pub enum ICU4XError {
+        // general errors
+        /// The error is not currently categorized as ICU4XError.
+        /// Please file a bug
+        UnknownError = 0,
+        /// An error arising from writing to a string
+        /// Typically found when not enough space is allocated
+        /// Most APIs that return a string may return this error
+        WriteableError = 1,
+
+        // Some input was out of bounds
+        OutOfBoundsError = 2,
+
+        // general data errors
+        // See DataError
+        DataMissingResourceKeyError = 10,
```

Comment:
Nit: I'd probably leave more room between sections. ICU4C utypes.h leaves `0x10000` between sections. I think that's probably too much, since ICU4X enums are 1-byte. Maybe at least start each section on a multiple of 16 and set these to hex values like `0x20`, `0x21`, etc.

Actually I'd kind-of prefer if the error enum were larger, like `u32`, so that we didn't have to think as hard about adding new entries.

In general, we should bikeshed this struct. It's fine to check in, but we should schedule this discussion for one of the meetings before the end of June.

Reply:
The error enum _is_ larger, it's a `usize`. `repr(C)` defaults to usize, `repr(Rust)` does not. Can leave a comment to that effect.

Reply:
Oh right, I was thinking of `#[make_ule]` on enum which limits it to `u8`. In this case, we should just use the larger values, then. I think probably 256 values per section is fine.

Reply:
Done.